### ISSUE TITLE
Feature suggestion: Mapper instances (alternative to using global mapper registration)

### ISSUE THREAD

**Issue by sh-at-cs** [dataclass-mapper/dataclass-mapper#33]
## Proposed feature

For reasons that will be outlined below, I think it would be useful if `create_mapper` returned "mapper instances" that can be used independently of `map_to` and its globally registered mappers, e.g.:

```python
from dataclasses import dataclass

from dataclass_mapper import create_mapper

@dataclass
class A:
    u: str

@dataclass
class B:
    u: str

a_to_b = create_mapper(A, B)
# type of a_to_b is roughly Callable[[A], B]

b = a_to_b(A("foo"))
```

### Better static type checking

Consider e.g.:

```python
from dataclasses import dataclass

from dataclass_mapper import create_mapper, map_to

@dataclass
class Inp1:
    u: str

@dataclass
class Inp2:
    u: str

@dataclass
class Out:
    u: str

create_mapper(Inp1, Out)

# Fails at runtime because we forgot to also create a mapping from Inp2 to Out
map_to(Inp2("foo"), Out)
```

In most projects, only failing at the point when `map_to` is called isn't so bad because this will show up in automated tests anyway. But even so, it's usually preferable to have such issues show up as early as possible, ideally during static type checking with Mypy or Pyright.

With "mapper instances" as demonstrated above, the issue would show up immediately: For people adhering to the naming scheme above, it would show up when writing the code, but even if someone makes a mistake there (e.g. by copy-pasting another mapper and forgetting to adjust the types), it will show up during static type checking at the latest, because the signature of the mapper instance only accepts arguments of type `Inp1`, unlike `map_to` which must accept anything.

### Allowing multiple different mappings between the same pair of types

Having mapper instances that are completely independent of each other would mean that someone can create different mappers between the same pair of types, to be used in different contexts. This probably isn't very useful in reality where you usually only need one kind of mapping between any two types, but who knows.

### Issues with registration via global state in general

These are all similar to criticisms that have been raised in response to Python's own [`singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch), which has a similar global registration mechanism:

- A module that performs the registration doesn't necessarily contain anything that needs to be imported from anywhere else, so it can happen that one forgets to import it and the registration will never happen. Confusion follows.
- It's easier/faster to find out where this or that type has been registered if you can just follow the trail of your IDE's "go to definition" feature than if it can come from anywhere in the imported modules and you have to know what you're looking for.

## Open questions

How would mapping of nested dataclasses be configured with this new API? Seems to me like no matter how this is handled, it will always diminish the static type checking advantages of the feature a bit (e.g. if it worked like `a_to_b = create_mapper(...); a_to_b.register_nested(NestA, NestB)`, you'd be back to only noticing errors at the time when the mapper is called...).

Thoughts?

**Comment by jakobkogler** [comment]
I'm not opposed to this idea. 

But I'm not completely sure how it would look like for nested conversions.
On of the nicer features of the global registration is, that you just need to define the individual mappings between classes, and the library figures all the things out by it's own.
```python
@dataclass
class InpChild:
    pass

@dataclass
class Inp:
    child: InpChild

@dataclass
class OutChild:
    pass

@dataclass
class Out:
    child: OutChild
```

Currently two simple mappings are enough:
```python
create_mapper(InpChild, OutChild)
create_mapper(Inp, Out)

inp = Inp(child=InpChild())
out = map_to(inp, Out)
```

With mapper instances, you would need to pass the mapper instances somehow. 
Not sure if I like this. 
```python
child_mapper = create_mapper(InpChild, OutChild)
main_mapper = create_mapper(Inp, Out, mapper_instance=[child_mapper])
```
Any other ideas?

**Comment by sh-at-cs** [comment]
Yeah, nested mappings are the major issue with this. But I think it looks similar to your last snippet in popular mapping libraries for other languages, [e.g. in Java's MapStruct it would be](https://mapstruct.org/documentation/stable/reference/html/#invoking-other-mappers): 
```java
@Mapper(uses=ChildMapper.class)
public interface MainMapper {
  ...
}
```
Which is the same general idea.

-----

By the way, a workaround (for users of dataclass-mapper, i.e. not a suggestion for how to build this into dataclass-mapper itself) that provides some of the type checking benefits of instance mappers is to just define a wrapper that returns `map_to` but pretends it has the appropriate static type signature:

```python
from typing import cast, TypeVar

from dataclass_mapper import create_mapper, map_to

S, T = TypeVar("S"), TypeVar("T")

def create_statically_typed_mapper(SourceCls: S, TargetCls: T, ...) -> Callable[[S], T]:
  create_mapper(SourceCls, TargetCls, ...)
  return cast(Callable[[S], T], map_to)
```

But as this still relies on global registration for nested mappings, it's not ideal.

-----

I also forgot to mention in the issue that another advantage of not using global registration would be allowing several different mappings between the same pair of types. But I don't know how useful that would be in reality, where most of the time you only want to have one such mapping anyway. Still, I'll add it to the list.

**Comment by jakobkogler** [comment]
I'll think about it in the next few days.

Another advantage is, that by having a mapper instance, you guarantee that the mapper was actually created.
Currently, if you define the mapper in a separate file that's not imported, the mapper is never created.

**Comment by jakobkogler** [comment]
I also thought about the problem of defining multiple mappings between two classes before.

I had planned to create something like mapper instances, however more in the sense of mapping groups. But than you obviously still have the problem of type safety.
